_posts/2021-03-09-strategies-for-small-focused-pull-requests.markdown

```diff
@@ -0,0 +1,122 @@
+
+When a large PR combines wide, shallow, abstract work with deep, narrow, concrete work, it requires the reviewer to shift between two different mindsets. You might consider breaking your PR into two: one containing the wide infrastructural work, and one containing the deep implementation work. This allows reviewers to focus on abstractions in one PR and user-facing details in the other. 
+
+Some examples of infrastructural changes that could be separated from implementation work:
+
+* We introduced a seam to the code in order to make room for our implementation. 
+* We updated a dependency to take advantage of a new feature. 
+* We refactored before we started our implementation.
+
+## Separating an already-large PR
+
+It's natural for PRs to grow large. [Optimism bias][optimism-bias] diminishes our ability to estimate work often resulting in more code changes than we expected. A feature seems like it won't take much work until you get deeper and find complexity in places you hadn't considered. There's a lot of uncertainty when you start working on a feature and we'd need to model the entire problem to completion to know what the PR was going to look like before we started. A PR seems like it will be small until suddenly...it isn't anymore.
+
+This is what usually prevents developers from separating PRs — by the time you recognize the PRs could be de-tangled, it seems like a lot of effort to de-tangle them. 
+
+When you've got a PR/branch that contains multiple lines of work and you want to separate them, [`git rebase`][git-rebase] is your best friend. Rebasing enables you to rename, reorder, combine, and separate commits. Use `git rebase` to group your commits into one set for each branch you want to extract, and submit a PR for each smaller set of changes. 
+
+Good commit hygiene makes it easier to rebase commits. Commit small units of work so that they can be re-ordered and grouped, and apply clear messages to each commit in case you need to move it. While you might not _always_ separate/rebase PR branches, you'll appreciate small commits with clear messages when you do. 
+
+## Small PRs start long before the work starts
+
+The size of a pull request can be influenced long before the PR is opened. Slice features small in your product backlog; make small commits along the way; combine small commits into small pull requests. Among other benefits, a focus on breaking work into small parts will make it easier to review your changes. 
+
+
+[draw-the-rest-of-the-owl]: https://knowyourmeme.com/memes/how-to-draw-an-owl
+[optimism-bias]: https://thedecisionlab.com/biases/optimism-bias/
+[git-rebase]: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase
+[walking-skeleton]: todo
+[react]: https://reactjs.org/
+[relay]: https://relay.dev/
+[typescript]: https://www.typescriptlang.org/
+[matt]: https://artsy.github.io/author/matt/
```

Comment:
@mzikherman I referenced a couple PRs from you in the article; let me know if you'd prefer me to link your name somewhere else!